Exceptions on develop-sling12 peregrine startup

These are Exceptions that I get when running `java -agentlib:jdwp=transport=dt_socket,address=8001,server=y,suspend=n -jar org.apache.sling.feature.launcher.jar -f com.peregrine-cms.sling.launchpad-12-SNAPSHOT-oak_tar_fds_far.far -D sling.runmodes=author,notshared,oak_tar_fds -p sling` on Windows with the branch [develop-sling12](https://github.com/headwirecom/peregrine-cms/tree/develop-sling12):
```W:\headwire\PeregrineCMS\percms\12>java -agentlib:jdwp=transport=dt_socket,address=8001,server=y,suspend=n -jar org.apache.sling.feature.launcher.jar -f com.peregrine-cms.sling.launchpad-12-SNAPSHOT-oak_tar_fds_far.far -D sling.runmodes=author,notshared,oak_tar_fds -p sling
Listening for transport dt_socket at address: 8001
[INFO]
[INFO] Apache Sling Application Launcher
[INFO] ---------------------------------
[INFO] Initializing...
[INFO] Artifact Repositories: [file:///C:/Users/reggi/.m2/repository, https://repo.maven.apache.org/maven2, https://repository.apache.org/content/groups/snapshots]
[INFO] Assembling provisioning model...
[INFO]
[INFO] Assembling launcher...
[INFO] Using 365 local artifacts, 0 cached artifacts, and 0 downloaded artifacts
[INFO]
[INFO] Starting launcher...
[INFO] Launcher Home: W:\headwire\PeregrineCMS\percms\12\sling
[INFO] Cache Directory: W:\headwire\PeregrineCMS\percms\12\launcher\cache
[INFO]
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.apache.felix.framework.ext.ClassPathExtenderFactory$DefaultClassLoaderExtender (file:/C:/Users/reggi/.m2/repository/org/apache/felix/org.apache.felix.
framework/6.0.3/org.apache.felix.framework-6.0.3.jar) to method java.net.URLClassLoader.addURL(java.net.URL)
WARNING: Please consider reporting this to the maintainers of org.apache.felix.framework.ext.ClassPathExtenderFactory$DefaultClassLoaderExtender
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Started Jetty 9.4.28.v20200408 at port(s) HTTP:8080 on context path / [minThreads=8,maxThreads=200,acceptors=1,selectors=4]
[INFO] Apachde Felix Http Whiteboard Service started
ERROR: bundle org.apache.sling.settings:1.4.0 (22)[org.apache.sling.settings.impl.SlingPropertiesPrinter(1)] :  Error during instantiation of the implementation object
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at org.apache.felix.scr.impl.inject.internal.ComponentConstructorImpl.newInstance(ComponentConstructorImpl.java:312)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:279)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:984)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:957)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:902)
        at org.apache.felix.framework.ServiceRegistrationImpl.getFactoryUnchecked(ServiceRegistrationImpl.java:348)
        at org.apache.felix.framework.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:248)
        at org.apache.felix.framework.ServiceRegistry.getService(ServiceRegistry.java:350)
        at org.apache.felix.framework.Felix.getService(Felix.java:3954)
        at org.apache.felix.framework.BundleContextImpl.getService(BundleContextImpl.java:450)
        at org.apache.felix.inventory.impl.webconsole.WebConsoleAdapter.addingService(WebConsoleAdapter.java:152)
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:871)
        at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
        at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183)
        at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:321)
        at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:264)
        at org.apache.felix.inventory.impl.webconsole.WebConsoleAdapter.<init>(WebConsoleAdapter.java:68)
        at org.apache.felix.inventory.impl.Activator.start(Activator.java:63)
        at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:698)
        at org.apache.felix.framework.Felix.activateBundle(Felix.java:2402)
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2308)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1539)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: No bundle context property 'sling.properties.url' provided
        at org.apache.sling.settings.impl.SlingPropertiesPrinter.<init>(SlingPropertiesPrinter.java:64)
        ... 30 more
INFO : org.apache.felix.webconsole.plugins.memoryusage (54): Storing Memory Dumps in W:\headwire\PeregrineCMS\percms\12\sling\framework\bundle54\data\dumps
INFO : org.apache.felix.webconsole.plugins.memoryusage (54): Setting Automatic Memory Dump Threshold to 0% for pools [CodeHeap 'non-nmethods', CodeHeap 'non-profiled nmethods', CodeHeap 'profi
led nmethods', Compressed Class Space, G1 Old Gen, Metaspace]
INFO : org.apache.felix.webconsole.plugins.memoryusage (54): Automatic Memory Dump cannot be set for pools [G1 Eden Space, G1 Survivor Space]
INFO : org.apache.felix.webconsole.plugins.memoryusage (54): Setting Automatic Memory Dump Interval to 21600 seconds
INFO : org.apache.felix.webconsole.plugins.memoryusage (54): Setting Automatic Memory Dump Threshold to 0% for pools [CodeHeap 'non-nmethods', CodeHeap 'non-profiled nmethods', CodeHeap 'profi
led nmethods', Compressed Class Space, G1 Old Gen, Metaspace]
INFO : org.apache.felix.webconsole.plugins.memoryusage (54): Automatic Memory Dump cannot be set for pools [G1 Eden Space, G1 Survivor Space]
INFO : org.apache.felix.webconsole.plugins.memoryusage (54): Setting Automatic Memory Dump Interval to 21600 seconds
INFO : org.apache.felix.webconsole.plugins.memoryusage (54): Storing Memory Dumps in W:\headwire\PeregrineCMS\percms\12\sling\framework\bundle54\data\dumps
wrz 23, 2020 2:02:37 PM org.apache.tika.config.InitializableProblemHandler$3 handleInitializableProblem
WARNING: TIFFImageWriter not loaded. tiff files will not be processed
See https://pdfbox.apache.org/2.0/dependencies.html#jai-image-io
for optional dependencies.
J2KImageReader not loaded. JPEG2000 files will not be processed.
See https://pdfbox.apache.org/2.0/dependencies.html#jai-image-io
for optional dependencies.

wrz 23, 2020 2:02:37 PM org.apache.tika.config.InitializableProblemHandler$3 handleInitializableProblem
WARNING: org.xerial's sqlite-jdbc is not loaded.
Please provide the jar on your classpath to parse sqlite files.
See tika-parsers/pom.xml for the correct version.
[DEBUG] [ServiceReference 421 from bundle 103 : org.apache.sling.jcr.davex:1.3.10 ref=[javax.servlet.Servlet] properties={objectClass=[javax.servlet.Servlet], osgi.http.whiteboard.context.sele
ct=(osgi.http.whiteboard.context.name=DavExAuthHttpContext), osgi.http.whiteboard.servlet.pattern=/server/*, service.bundleid=103, service.description=Sling JcrRemoting Servlet, service.id=421
, service.scope=singleton, service.vendor=The Apache Software Foundation, servlet.init.createAbsoluteURI=true, servlet.init.csrf-protection=disabled, servlet.init.protectedhandlers-config=org.
apache.jackrabbit.server.remoting.davex.AclRemoveHandler, servlet.init.resource-path-prefix=/server, sling.auth.requirements=-/server}] Ignoring unmatching Servlet service
ESAPI: WARNING: System property [org.owasp.esapi.opsteam] is not set
ESAPI: WARNING: System property [org.owasp.esapi.devteam] is not set
ESAPI: Attempting to load ESAPI.properties via file I/O.
ESAPI: Attempting to load ESAPI.properties as resource file via file I/O.
ESAPI: Not found in 'org.owasp.esapi.resources' directory or file not readable: W:\headwire\PeregrineCMS\percms\12\ESAPI.properties
ESAPI: Not found in SystemResource Directory/resourceDirectory: .esapi\ESAPI.properties
ESAPI: Not found in 'user.home' (C:\Users\reggi) directory: C:\Users\reggi\esapi\ESAPI.properties
ESAPI: Loading ESAPI.properties via file I/O failed. Exception was: java.io.FileNotFoundException
ESAPI: Attempting to load ESAPI.properties via the classpath.
ESAPI: SUCCESSFULLY LOADED ESAPI.properties via the CLASSPATH from '/ (root)' using class loader for DefaultSecurityConfiguration class!
ESAPI: SecurityConfiguration for Validator.ConfigurationFile.MultiValued not found in ESAPI.properties. Using default: false
ESAPI: Attempting to load validation.properties via file I/O.
ESAPI: Attempting to load validation.properties as resource file via file I/O.
ESAPI: Not found in 'org.owasp.esapi.resources' directory or file not readable: W:\headwire\PeregrineCMS\percms\12\validation.properties
ESAPI: Not found in SystemResource Directory/resourceDirectory: .esapi\validation.properties
ESAPI: Not found in 'user.home' (C:\Users\reggi) directory: C:\Users\reggi\esapi\validation.properties
ESAPI: Loading validation.properties via file I/O failed.
ESAPI: Attempting to load validation.properties via the classpath.
ESAPI: SUCCESSFULLY LOADED validation.properties via the CLASSPATH from '/ (root)' using class loader for DefaultSecurityConfiguration class!
Warning: Nashorn engine is planned to be removed from a future JDK release
Warning: Nashorn engine is planned to be removed from a future JDK release
Warning: Nashorn engine is planned to be removed from a future JDK release
Warning: Nashorn engine is planned to be removed from a future JDK release
Warning: Nashorn engine is planned to be removed from a future JDK release
[INFO] Framework started
23.09.2020 14:34:07.167 *INFO* [OsgiInstallerImpl] org.apache.sling.audit.osgi.installer Installed configuration org.apache.sling.jcr.repoinit.RepositoryInitializer~peregrine from resource Tas
kResource(url=jcrinstall:/apps/runmodes/config/org.apache.sling.jcr.repoinit.RepositoryInitializer-peregrine.config, entity=config:org.apache.sling.jcr.repoinit.RepositoryInitializer~peregrine
, state=INSTALL, attributes=[org.apache.sling.installer.api.tasks.ResourceTransformer=:21:, service.factoryPid=org.apache.sling.jcr.repoinit.RepositoryInitializer, service.pid=peregrine], dige
st=c5c7279b728b00c506832d23642f265a)
23.09.2020 14:34:07.177 *INFO* [OsgiInstallerImpl] org.apache.sling.audit.osgi.installer Installed configuration org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~peregrin
e from resource TaskResource(url=jcrinstall:/apps/runmodes/config/org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended-peregrine, entity=config:org.apache.sling.serviceuserma
pping.impl.ServiceUserMapperImpl.amended~peregrine, state=INSTALL, attributes=[org.apache.sling.installer.api.tasks.ResourceTransformer=:21:, service.factoryPid=org.apache.sling.serviceusermap
ping.impl.ServiceUserMapperImpl.amended, service.pid=peregrine], digest=e97736490720f80c1e05695b11f406ec)
ERROR: bundle com.peregrine-cms.replication.core:1.0.0.SNAPSHOT (193)[com.peregrine.admin.replication.impl.ReplicationMixinNodeTypesOverride(458)] : The activate method has thrown an exception

java.lang.IllegalArgumentException: Could not obtain Resource Resolver -> setup failed
        at com.peregrine.admin.replication.impl.ReplicationMixinNodeTypesOverride.setup(ReplicationMixinNodeTypesOverride.java:97)
        at com.peregrine.admin.replication.impl.ReplicationMixinNodeTypesOverride.activate(ReplicationMixinNodeTypesOverride.java:62)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:242)
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41)
        at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:678)
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:524)
        at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:318)
        at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:308)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:342)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:984)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:957)
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:766)
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:667)
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:433)
        at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:666)
        at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:353)
        at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:553)
        at org.apache.felix.scr.impl.Activator.access$200(Activator.java:69)
        at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:424)
        at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196)
        at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169)
        at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49)
        at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:488)
        at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:420)
        at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232)
        at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:450)
        at org.apache.felix.framework.EventDispatcher.invokeBundleListenerCallback(EventDispatcher.java:915)
        at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:834)
        at org.apache.felix.framework.EventDispatcher.fireBundleEvent(EventDispatcher.java:516)
        at org.apache.felix.framework.Felix.fireBundleEvent(Felix.java:4817)
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2336)
        at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998)
        at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984)
        at org.apache.sling.installer.core.impl.tasks.BundleStartTask.execute(BundleStartTask.java:97)
        at org.apache.sling.installer.core.impl.OsgiInstallerImpl.doExecuteTasks(OsgiInstallerImpl.java:918)
        at org.apache.sling.installer.core.impl.OsgiInstallerImpl.executeTasks(OsgiInstallerImpl.java:755)
        at org.apache.sling.installer.core.impl.OsgiInstallerImpl.run(OsgiInstallerImpl.java:304)
        at java.base/java.lang.Thread.run(Thread.java:834)```

Regarding the other one. 

ERROR: bundle org.apache.sling.settings:1.4.0 (22)[org.apache.sling.settings.impl.SlingPropertiesPrinter(1)] :  Error during instantiation of the implementation object
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at org.apache.felix.scr.impl.inject.internal.ComponentConstructorImpl.newInstance(ComponentConstructorImpl.java:312)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:279)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:984)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:957)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:902)
        at org.apache.felix.framework.ServiceRegistrationImpl.getFactoryUnchecked(ServiceRegistrationImpl.java:348)
        at org.apache.felix.framework.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:248)
        at org.apache.felix.framework.ServiceRegistry.getService(ServiceRegistry.java:350)
        at org.apache.felix.framework.Felix.getService(Felix.java:3954)
        at org.apache.felix.framework.BundleContextImpl.getService(BundleContextImpl.java:450)
        at org.apache.felix.inventory.impl.webconsole.WebConsoleAdapter.addingService(WebConsoleAdapter.java:152)
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:871)
        at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
        at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183)
        at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:321)
        at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:264)
        at org.apache.felix.inventory.impl.webconsole.WebConsoleAdapter.<init>(WebConsoleAdapter.java:68)
        at org.apache.felix.inventory.impl.Activator.start(Activator.java:63)
        at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:698)
        at org.apache.felix.framework.Felix.activateBundle(Felix.java:2402)
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2308)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1539)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: No bundle context property 'sling.properties.url' provided
        at org.apache.sling.settings.impl.SlingPropertiesPrinter.<init>(SlingPropertiesPrinter.java:64)
        ... 30 more

My understanding is that Sling has a fix for it coming....
https://issues.apache.org/jira/browse/SLING-9648

--- a/platform/base/core/src/main/java/com/peregrine/adaption/impl/PerPageManagerImpl.java
+++ b/platform/base/core/src/main/java/com/peregrine/adaption/impl/PerPageManagerImpl.java
@@ -31,6 +31,8 @@ import org.apache.sling.api.resource.ModifiableValueMap;
 import org.apache.sling.api.resource.PersistenceException;
 import org.apache.sling.api.resource.Resource;
 import org.apache.sling.api.resource.ResourceResolver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.Calendar;
 
@@ -47,6 +49,8 @@ public class PerPageManagerImpl
 {
     public static final String RESOURCE_RESOLVER_MUST_BE_PROVIDED = "Resource Resolver must be provided";
     public static final String RESOURCE_MUST_BE_PROVIDED = "Resource must be provided";
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
     /** Resource Resolver that backs that Page Manager **/
     private ResourceResolver resourceResolver;
 
@@ -92,6 +96,7 @@ public class PerPageManagerImpl
                 resourceResolver.commit();
             } catch(PersistenceException e) {
                 //AS TODO: Log exception
+                logger.error("could not clear replication properties", e);
             }
             if(!shallow) {
                 for(PerPage child : page.listChildren()) {

--- a/platform/base/core/src/main/java/com/peregrine/adaption/impl/PerPageManagerImpl.java
+++ b/platform/base/core/src/main/java/com/peregrine/adaption/impl/PerPageManagerImpl.java
@@ -95,7 +95,6 @@ public class PerPageManagerImpl
             try {
                 resourceResolver.commit();
             } catch(PersistenceException e) {
-                //AS TODO: Log exception
                 logger.error("could not clear replication properties", e);
             }
             if(!shallow) {

--- a/platform/base/ui.apps/src/main/content/jcr_root/apps/runmodes/config/com.peregrine.admin.replication.impl.ReplicationMixinNodeTypesOverride.xml
+++ b/platform/base/ui.apps/src/main/content/jcr_root/apps/runmodes/config/com.peregrine.admin.replication.impl.ReplicationMixinNodeTypesOverride.xml
@@ -1,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  #%L
+  Multi Package Update - UI Apps
+  %%
+  Copyright (C) 2017 headwire inc.
+  %%
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing,
+  software distributed under the License is distributed on an
+  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  KIND, either express or implied.  See the License for the
+  specific language governing permissions and limitations
+  under the License.
+  #L%
+  -->
+<jcr:root xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:sling="http://sling.apache.org/jcr/sling/1.0"
+    jcr:primaryType="sling:OsgiConfig"
+    replicationNodeTypes="[per:AssetContent,per:PageContent]"
+/>

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/ReplicationUtil.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/ReplicationUtil.java
@@ -1,6 +1,5 @@
 package com.peregrine.admin.replication;
 
-import com.peregrine.commons.ResourceUtils;
 import org.apache.sling.api.resource.*;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -100,12 +99,13 @@ public class ReplicationUtil {
                 ModifiableValueMap sourceProperties = getModifiableProperties(source, false);
                 if (sourceProperties != null) {
                     Calendar replicated = Calendar.getInstance();
+                    final ResourceResolver resourceResolver = source.getResourceResolver();
                     if (DISTRIBUTION_PENDING.equals(targetPath)) {
                         // Note for remote replication use-case
                         // updateReplicationProperties will be called twice. The first time will include a resource
                         //   obtained from the user initiating resource publishing. In this case targetPath will be "distribution pending"
                         // TODO: CR add replication status to the per:Replication mixin such that these inferences are not needed.
-                        sourceProperties.put(PER_REPLICATED_BY, source.getResourceResolver().getUserID());
+                        sourceProperties.put(PER_REPLICATED_BY, resourceResolver.getUserID());
                     }
                     sourceProperties.put(PER_REPLICATED, replicated);
                     LOGGER.trace("Updated Source Replication Properties");
@@ -120,7 +120,7 @@ public class ReplicationUtil {
                         ensureMixin(target);
                         try {
                             ModifiableValueMap targetProperties = getModifiableProperties(target, false);
-                            String userId = source.getResourceResolver().getUserID();
+                            String userId = resourceResolver.getUserID();
                             LOGGER.trace("Replication User Id: '{}' for target: '{}'", userId, target.getPath());
                             // TODO: Refactor duplicated code
                             if (DISTRIBUTION_PENDING.equals(targetPath)) {
@@ -141,7 +141,8 @@ public class ReplicationUtil {
                             throw e;
                         }
                     }
-                    commitChanges(source);
+
+                    refreshAndCommit(resourceResolver);
                 } else {
                     LOGGER.debug("Source: '{}' is not writable -> ignored", source);
                 }
@@ -149,11 +150,11 @@ public class ReplicationUtil {
         }
     }
 
-    private static void commitChanges(Resource resource) {
-        resource.getResourceResolver().refresh();
+    private static void refreshAndCommit(final ResourceResolver resourceResolver) {
+        resourceResolver.refresh();
         try {
-            resource.getResourceResolver().commit();
-        } catch (PersistenceException e) {
+            resourceResolver.commit();
+        } catch (final PersistenceException e) {
             LOGGER.error("could not commit replication property changes", e);
         }
     }

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/ReplicationUtil.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/ReplicationUtil.java
@@ -101,8 +101,9 @@ public class ReplicationUtil {
                 if (sourceProperties != null) {
                     Calendar replicated = Calendar.getInstance();
                     if (DISTRIBUTION_PENDING.equals(targetPath)) {
+                        // Note for remote replication use-case
                         // updateReplicationProperties will be called twice. The first time will include a resource
-                        // obtained from the user initiating resource publishing. In this case targetPath will be "distribution pending"
+                        //   obtained from the user initiating resource publishing. In this case targetPath will be "distribution pending"
                         // TODO: CR add replication status to the per:Replication mixin such that these inferences are not needed.
                         sourceProperties.put(PER_REPLICATED_BY, source.getResourceResolver().getUserID());
                     }
@@ -140,6 +141,7 @@ public class ReplicationUtil {
                             throw e;
                         }
                     }
+                    commitChanges(source);
                 } else {
                     LOGGER.debug("Source: '{}' is not writable -> ignored", source);
                 }
@@ -147,7 +149,14 @@ public class ReplicationUtil {
         }
     }
 
-
+    private static void commitChanges(Resource resource) {
+        resource.getResourceResolver().refresh();
+        try {
+            resource.getResourceResolver().commit();
+        } catch (PersistenceException e) {
+            LOGGER.error("could not commit replication property changes", e);
+        }
+    }
 
 //    /**
 //     * References under /content for replication

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/BaseFileReplicationService.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/BaseFileReplicationService.java
@@ -66,13 +66,15 @@ public abstract class BaseFileReplicationService
     private static final List<Pattern> NAME_PATTERNS = new ArrayList<>();
     // List of all resources that are excluded from handling
     static final List<String> EXCLUDED_RESOURCES = new ArrayList<>();
+    public static String DATA_PATTERN = ".*(\\.data\\.json|\\.html)";
 
     private static final String EXTENSION_NAME_MUST_BE_PROVIDED = "Extension Name must be provided";
     private static final String EXTENSION_TYPES_MUST_BE_PROVIDED = "Extension Types must be provided";
     public static final String RENDERING_OF_ASSET_FAILED = "Rendering of Asset failed";
 
     static {
-        NAME_PATTERNS.add(Pattern.compile(".*\\.data\\.json"));
+        NAME_PATTERNS.add(Pattern.compile(DATA_PATTERN));
+//        NAME_PATTERNS.add(Pattern.compile(HTML_PATTERN));
         EXCLUDED_RESOURCES.add(JCR_CONTENT);
         EXCLUDED_RESOURCES.add(RENDITIONS);
     }

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/DistributionEventHandlerService.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/DistributionEventHandlerService.java
@@ -86,13 +86,8 @@ public class DistributionEventHandlerService implements EventHandler {
                 log.info("properties for {} were updated by dist event handler.",path);
                 updateReplicationProperties(resource, replicationRef, null);
             }
-
-            resourceResolver.commit();
-
-        } catch (LoginException | PersistenceException e) {
+        } catch (LoginException e) {
             log.error("Failed to update per:Replication properties", e);
-        } finally {
-            resourceResolver.close();
         }
     }
 

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/DistributionReplicationService.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/DistributionReplicationService.java
@@ -202,11 +202,7 @@ public class DistributionReplicationService
                     // In order to make it possible to have the correct user set and 'Replicated By' we need to set it here and now
                     updateReplicationProperties(resource, DISTRIBUTION_PENDING, null);
                 }
-                try {
-                    resourceResolver.commit();
-                } catch(PersistenceException e) {
-                    throw new ReplicationException("Could not set Replication User before distribution", e);
-                }
+
                 if(distributor != null) {
                     if (activate) {
                         // first deactivate page content so deleted or moved content is cleared

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
@@ -42,11 +42,9 @@ import java.io.File;
 import java.io.FileOutputStream;
 import java.io.FileWriter;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
 import java.util.Map.Entry;
+import java.util.function.Function;
 import java.util.regex.Pattern;
 
 import static com.peregrine.admin.replication.ReplicationUtil.updateReplicationProperties;
@@ -59,6 +57,8 @@ import static com.peregrine.commons.util.PerUtil.isNotEmpty;
 import static com.peregrine.commons.util.PerUtil.splitIntoMap;
 import static com.peregrine.commons.util.PerUtil.splitIntoProperties;
 import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
 
 /**
  * This class replicates resources to a local file system folder
@@ -277,32 +277,42 @@ public class LocalFileSystemReplicationService
 
     @Override
     String storeRendering(Resource resource, String extension, String content) throws ReplicationException {
-        File renderingFile = createRenderingFile(resource, extension);
-        try {
-            FileWriter fileWriter = new FileWriter(renderingFile);
-            fileWriter.append(content);
-            fileWriter.close();
-        } catch(IOException e) {
-            throw new ReplicationException(String.format(FAILED_TO_STORE_RENDERING, renderingFile.getAbsolutePath()), e);
+        return storeRendering(resource, extension, file -> {
+            try (final FileWriter fileWriter = new FileWriter(file)) {
+                fileWriter.append(content);
+            } catch (final IOException e) {
+                return new ReplicationException(String.format(FAILED_TO_STORE_RENDERING, file.getAbsolutePath()), e);
+            }
+
+            return null;
+        });
+    }
+
+    private String storeRendering(final Resource resource, final String extension, final Function<File, ReplicationException> fileConsumer)
+            throws ReplicationException {
+        final File renderingFile = createRenderingFile(resource, extension);
+        final ReplicationException exception = fileConsumer.apply(renderingFile);
+        if (nonNull(exception)) {
+            throw exception;
         }
-        Resource propertiesRes = resource.getName().equals(JCR_CONTENT) ? resource : resource.getChild(JCR_CONTENT);
-        updateReplicationProperties(propertiesRes, LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath(), null);
-        return LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath();
+
+        final Resource propertiesRes = PerUtil.getJcrContent(resource);
+        final String localFileSystemPath = LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath();
+        updateReplicationProperties(propertiesRes, localFileSystemPath, null);
+        return localFileSystemPath;
     }
 
     @Override
     String storeRendering(Resource resource, String extension, byte[] content) throws ReplicationException {
-        File renderingFile = createRenderingFile(resource, extension);
-        try {
-            FileOutputStream fileOutputStream = new FileOutputStream(renderingFile);
-            fileOutputStream.write(content);
-            fileOutputStream.close();
-        } catch(IOException e) {
-            throw new ReplicationException(String.format(CANNOT_WRITE_RENDERING, renderingFile.getAbsolutePath()), e);
-        }
-        Resource propertiesRes = resource.getName().equals(JCR_CONTENT) ? resource : resource.getChild(JCR_CONTENT);
-        updateReplicationProperties(propertiesRes, LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath(), null);
-        return LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath();
+        return storeRendering(resource, extension, file -> {
+            try (final FileOutputStream fileOutputStream = new FileOutputStream(file)) {
+                fileOutputStream.write(content);
+            } catch (final IOException e) {
+                return new ReplicationException(String.format(CANNOT_WRITE_RENDERING, file.getAbsolutePath()), e);
+            }
+
+            return null;
+        });
     }
 
     @Override

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
@@ -25,7 +25,6 @@ package com.peregrine.admin.replication.impl;
  * #L%
  */
 
-import com.peregrine.commons.util.PerUtil;
 import com.peregrine.render.RenderService;
 import com.peregrine.replication.ReferenceLister;
 import com.peregrine.replication.Replication;
@@ -52,12 +51,10 @@ import static com.peregrine.commons.Chars._SCORE;
 import static com.peregrine.commons.ResourceUtils.jcrNameToFileName;
 import static com.peregrine.commons.util.PerConstants.JCR_CONTENT;
 import static com.peregrine.commons.util.PerConstants.SLASH;
-import static com.peregrine.commons.util.PerUtil.intoList;
-import static com.peregrine.commons.util.PerUtil.isNotEmpty;
-import static com.peregrine.commons.util.PerUtil.splitIntoMap;
-import static com.peregrine.commons.util.PerUtil.splitIntoProperties;
+import static com.peregrine.commons.util.PerUtil.*;
 import static java.util.Objects.isNull;
 import static java.util.Objects.nonNull;
+import static org.apache.commons.lang3.StringUtils.EMPTY;
 import static org.apache.commons.lang3.StringUtils.isNotBlank;
 
 /**
@@ -296,9 +293,8 @@ public class LocalFileSystemReplicationService
             throw exception;
         }
 
-        final Resource propertiesRes = PerUtil.getJcrContent(resource);
         final String localFileSystemPath = LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath();
-        updateReplicationProperties(propertiesRes, localFileSystemPath, null);
+        updateReplicationProperties(getJcrContent(resource), localFileSystemPath, null);
         return localFileSystemPath;
     }
 
@@ -325,16 +321,17 @@ public class LocalFileSystemReplicationService
 
         final File[] filesToBeDeletedFiles = directory.listFiles(
                 file -> {
-                    if (isFolder && file.isDirectory() && file.getName().equals(resourceName)) {
+                    final String name = file.getName();
+                    if (isFolder && file.isDirectory() && name.equals(resourceName)) {
                         return true;
                     }
 
                     if (isNull(namePattern)) {
-                        return file.getName().startsWith(resourceName);
+                        return name.startsWith(resourceName);
                     }
 
                     for (final Pattern pattern : namePattern) {
-                        if (pattern.matcher(file.getName()).matches()) {
+                        if (pattern.matcher(name).matches()) {
                             return true;
                         }
                     }
@@ -342,15 +339,17 @@ public class LocalFileSystemReplicationService
                     return false;
                 }
         );
-        if(filesToBeDeletedFiles != null) {
-            for(File toBeDeleted : filesToBeDeletedFiles) {
-                log.trace("Delete File: '{}'", toBeDeleted.getAbsolutePath());
-                if (!deleteFile(toBeDeleted)) {
-                    throw new ReplicationException(String.format(FAILED_TO_DELETE_FILE, toBeDeleted.getAbsolutePath()));
-                }
-                Resource propertiesRes = resource.getName().equals(JCR_CONTENT) ? resource : resource.getChild(JCR_CONTENT);
-                updateReplicationProperties(propertiesRes, "", null);
+        if (isNull(filesToBeDeletedFiles)) {
+            return;
+        }
+
+        for (final File toBeDeleted : filesToBeDeletedFiles) {
+            log.trace("Delete File: '{}'", toBeDeleted.getAbsolutePath());
+            if (!deleteFile(toBeDeleted)) {
+                throw new ReplicationException(String.format(FAILED_TO_DELETE_FILE, toBeDeleted.getAbsolutePath()));
             }
+
+            updateReplicationProperties(getJcrContent(resource), EMPTY, null);
         }
     }
 

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
@@ -324,9 +324,7 @@ public class LocalFileSystemReplicationService
                     }
 
                     for (final Pattern pattern : namePattern) {
-                        if (pattern.matcher(file.getName()).matches()) {
-                            return true;
-                        }
+                        return pattern.matcher(file.getName()).matches() && file.getName().startsWith(resourceName);
                     }
 
                     return false;

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
@@ -50,11 +50,15 @@ import java.util.Map.Entry;
 import java.util.regex.Pattern;
 
 import static com.peregrine.admin.replication.ReplicationUtil.updateReplicationProperties;
+import static com.peregrine.commons.Chars._SCORE;
+import static com.peregrine.commons.ResourceUtils.jcrNameToFileName;
 import static com.peregrine.commons.util.PerConstants.JCR_CONTENT;
+import static com.peregrine.commons.util.PerConstants.SLASH;
 import static com.peregrine.commons.util.PerUtil.intoList;
 import static com.peregrine.commons.util.PerUtil.isNotEmpty;
 import static com.peregrine.commons.util.PerUtil.splitIntoMap;
 import static com.peregrine.commons.util.PerUtil.splitIntoProperties;
+import static java.util.Objects.isNull;
 
 /**
  * This class replicates resources to a local file system folder

--- a/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
+++ b/platform/replication/core/src/main/java/com/peregrine/admin/replication/impl/LocalFileSystemReplicationService.java
@@ -49,11 +49,12 @@ import java.util.Map;
 import java.util.Map.Entry;
 import java.util.regex.Pattern;
 
-import static com.peregrine.commons.ResourceUtils.jcrNameToFileName;
-import static com.peregrine.commons.Strings.SLASH;
-import static com.peregrine.commons.Strings._SCORE;
-import static com.peregrine.commons.util.PerUtil.*;
-import static java.util.Objects.isNull;
+import static com.peregrine.admin.replication.ReplicationUtil.updateReplicationProperties;
+import static com.peregrine.commons.util.PerConstants.JCR_CONTENT;
+import static com.peregrine.commons.util.PerUtil.intoList;
+import static com.peregrine.commons.util.PerUtil.isNotEmpty;
+import static com.peregrine.commons.util.PerUtil.splitIntoMap;
+import static com.peregrine.commons.util.PerUtil.splitIntoProperties;
 
 /**
  * This class replicates resources to a local file system folder
@@ -280,6 +281,8 @@ public class LocalFileSystemReplicationService
         } catch(IOException e) {
             throw new ReplicationException(String.format(FAILED_TO_STORE_RENDERING, renderingFile.getAbsolutePath()), e);
         }
+        Resource propertiesRes = resource.getName().equals(JCR_CONTENT) ? resource : resource.getChild(JCR_CONTENT);
+        updateReplicationProperties(propertiesRes, LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath(), null);
         return LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath();
     }
 
@@ -293,6 +296,8 @@ public class LocalFileSystemReplicationService
         } catch(IOException e) {
             throw new ReplicationException(String.format(CANNOT_WRITE_RENDERING, renderingFile.getAbsolutePath()), e);
         }
+        Resource propertiesRes = resource.getName().equals(JCR_CONTENT) ? resource : resource.getChild(JCR_CONTENT);
+        updateReplicationProperties(propertiesRes, LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath(), null);
         return LOCAL_FILE_SYSTEM + renderingFile.getAbsolutePath();
     }
 
@@ -329,7 +334,8 @@ public class LocalFileSystemReplicationService
                 if (!deleteFile(toBeDeleted)) {
                     throw new ReplicationException(String.format(FAILED_TO_DELETE_FILE, toBeDeleted.getAbsolutePath()));
                 }
-
+                Resource propertiesRes = resource.getName().equals(JCR_CONTENT) ? resource : resource.getChild(JCR_CONTENT);
+                updateReplicationProperties(propertiesRes, "", null);
             }
         }
     }

--- a/slingjunit.parent/core/pom.xml
+++ b/slingjunit.parent/core/pom.xml
@@ -118,11 +118,15 @@
       <groupId>org.osgi</groupId>
       <artifactId>org.osgi.compendium</artifactId>
     </dependency>
-
     <dependency>
       <groupId>org.apache.httpcomponents</groupId>
       <artifactId>httpclient</artifactId>
       <version>4.5.12</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.jackrabbit</groupId>
+      <artifactId>oak-core-spi</artifactId>
+      <version>1.32.0</version>
     </dependency>
   </dependencies>
 

--- a/slingjunit.parent/core/src/main/java/com/peregrine/slingjunit/ReplicationTestBase.java
+++ b/slingjunit.parent/core/src/main/java/com/peregrine/slingjunit/ReplicationTestBase.java
@@ -1,0 +1,55 @@
+package com.peregrine.slingjunit;
+
+import com.peregrine.replication.Replication;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.sling.api.resource.Resource;
+
+
+import java.io.IOException;
+import java.util.Calendar;
+
+import static java.util.Objects.isNull;
+import static org.junit.Assert.*;
+
+public abstract class ReplicationTestBase  {
+
+    private static String PUBLISH_DOMAIN = "http://localhost:8180";
+
+
+    protected Calendar beforeTime;
+    protected void setPublishDomain(String publishDomain) {
+        this.publishDomain = publishDomain;
+    }
+
+    private String publishDomain;
+
+    protected void assertPublishedStatus(String path, int status) {
+        String domain = ! isNull(this.publishDomain) ? this.publishDomain : PUBLISH_DOMAIN;
+        HttpUriRequest request = new HttpGet( domain + path);
+
+        int retries = 10;
+        int attemptStatus = 0;
+        try {
+            while (attemptStatus != status && retries-- > 0) {
+                Thread.sleep(1000);
+                HttpResponse httpResponse = HttpClientBuilder.create().build().execute( request );
+                attemptStatus = httpResponse.getStatusLine().getStatusCode();
+            }
+            assertEquals(status, attemptStatus );
+        } catch (IOException | InterruptedException e) {
+            fail(e.getMessage());
+        }
+    }
+
+
+    protected void deactivateResource(String path, Resource resource, Replication replication) {
+        try {
+            replication.deactivate(resource);
+        } catch (Replication.ReplicationException e) {
+            fail(e.getMessage());
+        }
+    }
+}

--- a/slingjunit.parent/core/src/main/java/com/peregrine/slingjunit/author/RemoteReplAuthorJTest.java
+++ b/slingjunit.parent/core/src/main/java/com/peregrine/slingjunit/author/RemoteReplAuthorJTest.java
@@ -3,10 +3,7 @@ package com.peregrine.slingjunit.author;
 
 import com.peregrine.adaption.PerReplicable;
 import com.peregrine.replication.Replication;
-import org.apache.http.HttpResponse;
-import org.apache.http.client.methods.HttpGet;
-import org.apache.http.client.methods.HttpUriRequest;
-import org.apache.http.impl.client.HttpClientBuilder;
+import com.peregrine.slingjunit.ReplicationTestBase;
 import org.apache.sling.api.resource.LoginException;
 import org.apache.sling.api.resource.Resource;
 import org.apache.sling.api.resource.ResourceResolver;
@@ -35,7 +32,7 @@ import static org.apache.jackrabbit.JcrConstants.JCR_LASTMODIFIED;
 import static org.junit.Assert.*;
 
 @RunWith(SlingAnnotationsTestRunner.class)
-public class RemoteReplAuthorJTest {
+public class RemoteReplAuthorJTest extends ReplicationTestBase {
 
     @TestReference
     private SlingSettingsService settingsService;
@@ -45,14 +42,11 @@ public class RemoteReplAuthorJTest {
     @TestReference
     private ResourceResolverFactory resolverFactory;
 
+    @TestReference(name="remote")
+    Replication replication;
 
     @TestReference
     private ConfigurationAdmin configAdmin;
-
-    @TestReference(name="remote")
-    private Replication remoteReplication;
-
-    private Calendar beforeTime;
 
     private static String MAPPER_DISTRIBUTION_EVENT_PID = "org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~distributionEventHandler";
     private static String[] DIST_EVENT_VALUES = {
@@ -67,20 +61,19 @@ public class RemoteReplAuthorJTest {
             "com.peregrine-cms.distribution:test-distribution=distribution-agent-user"
     };
     private static String USER_MAPPING = "user.mapping";
-    private static String STELLA_PNG = "/content/example/assets/images/Stella.png";
+    public static String STELLA_PNG = "/content/example/assets/images/Stella.png";
     private static String CONTACT_PATH = "/content/example/pages/contact";
     private static String COMPONENT_PATH= "/content/example/pages/contact/jcr:content/n3736dc36-9cc3-49d7-a7d4-bf4d94e0ea2f";
     private Resource stellaImgRes;
-    private static String PUBLISH_DOMAIN = "http://localhost:8180";
+
 
     @Before
     public void setup(){
         try {
             beforeTime = Calendar.getInstance();
             adminResourceResolver = resolverFactory.getAdministrativeResourceResolver(null);
-            assertNotNull(remoteReplication);
+            assertNotNull(replication);
             stellaImgRes = adminResourceResolver.getResource(STELLA_PNG);
-
         } catch (LoginException e) {
             fail(e.getMessage());
         }
@@ -134,13 +127,13 @@ public class RemoteReplAuthorJTest {
 
     @Test
     public void replicateOneAsset() {
-        deactivateResource(STELLA_PNG);
-        assertStatusOnPublish(STELLA_PNG, 404);
+        deactivateResource(STELLA_PNG, stellaImgRes, replication);
+        assertPublishedStatus(STELLA_PNG, 404);
         PerReplicable stellaImgRepl = stellaImgRes.adaptTo(PerReplicable.class);
         assertFalse(stellaImgRepl.isReplicated());
         try {
-            remoteReplication.replicate(stellaImgRes, true);
-            assertStatusOnPublish(STELLA_PNG, 200);
+            replication.replicate(stellaImgRes, true);
+            assertPublishedStatus(STELLA_PNG, 200);
             assertTrue(stellaImgRepl.isReplicated());
 
             stellaImgRepl.getModifiableProperties().put(JCR_LASTMODIFIED, Calendar.getInstance());
@@ -150,8 +143,8 @@ public class RemoteReplAuthorJTest {
         } catch (Replication.ReplicationException e) {
             fail(e.getMessage());
         }
-        deactivateResource(STELLA_PNG);
-        assertStatusOnPublish(STELLA_PNG, 404);
+        deactivateResource(STELLA_PNG, stellaImgRes, replication);
+        assertPublishedStatus(STELLA_PNG, 404);
 
         stellaImgRepl = adminResourceResolver.getResource(STELLA_PNG).adaptTo(PerReplicable.class);
         assertFalse(stellaImgRepl.isReplicated());
@@ -190,33 +183,6 @@ public class RemoteReplAuthorJTest {
         } catch (IOException e) {
             fail(e.getMessage());
         }
-    }
-
-
-    private void assertStatusOnPublish(String path, int status) {
-        HttpUriRequest request = new HttpGet( PUBLISH_DOMAIN+path);
-
-        int retries = 10;
-        int attemptStatus = 0;
-        try {
-            while (attemptStatus != status && retries-- > 0) {
-                Thread.sleep(1000);
-                HttpResponse httpResponse = HttpClientBuilder.create().build().execute( request );
-                attemptStatus = httpResponse.getStatusLine().getStatusCode();
-            }
-            assertEquals(status, attemptStatus );
-        } catch (IOException | InterruptedException e) {
-            fail(e.getMessage());
-        }
-    }
-
-    private void deactivateResource(String path) {
-        try {
-            remoteReplication.deactivate(stellaImgRes);
-        } catch (Replication.ReplicationException e) {
-            fail(e.getMessage());
-        }
-        assertStatusOnPublish(path, 404);
     }
 
 }

--- a/slingjunit.parent/core/src/main/java/com/peregrine/slingjunit/localFS/LocalFSJTest.java
+++ b/slingjunit.parent/core/src/main/java/com/peregrine/slingjunit/localFS/LocalFSJTest.java
@@ -35,8 +35,14 @@ public class LocalFSJTest extends ReplicationTestBase {
 //    Otherwise, since the following line is non-deterministic, it could inject any of the replication services configured.
     @TestReference
     private Replication replication;
+    public static String INDEX = "/content/example/pages/index";
+    public static String CONTACT = "/content/example/pages/contact";
+    public static String JSON_DATA = ".data.json";
+    public static String HTML = ".html";
 
     private Resource stellaImgRes;
+    private Resource indexPageRes;
+    private Resource contactPageRes;
     private static String STATIC_HOME = "./sling/staticreplication";
 
     @Before
@@ -45,6 +51,8 @@ public class LocalFSJTest extends ReplicationTestBase {
             beforeTime = Calendar.getInstance();
             adminResourceResolver = resolverFactory.getAdministrativeResourceResolver(null);
             stellaImgRes = adminResourceResolver.getResource(STELLA_PNG);
+            indexPageRes = this.adminResourceResolver.getResource(INDEX);
+            contactPageRes = this.adminResourceResolver.getResource(CONTACT);
         } catch (LoginException e) {
             fail(e.getMessage());
         }
@@ -52,6 +60,9 @@ public class LocalFSJTest extends ReplicationTestBase {
 
     @After
     public void cleanup(){
+        deactivateResource(STELLA_PNG, stellaImgRes, replication);
+        deactivateResource(INDEX, indexPageRes, replication);
+        deactivateResource(INDEX, contactPageRes, replication);
         adminResourceResolver.close();
         adminResourceResolver = null;
     }
@@ -97,6 +108,42 @@ public class LocalFSJTest extends ReplicationTestBase {
         stellaImgRepl = adminResourceResolver.getResource(STELLA_PNG).adaptTo(PerReplicable.class);
         assertFalse(stellaImgRepl.isReplicated());
         assertFalse(stellaImgRepl.isStale());
+    }
+
+    @Test
+    public void replicatePage(){
+        assertFileDoesNotExist(STATIC_HOME+ INDEX + HTML);
+        assertFileDoesNotExist(STATIC_HOME+ CONTACT + HTML);
+        assertFileDoesNotExist(STATIC_HOME+ INDEX + JSON_DATA);
+        assertFileDoesNotExist(STATIC_HOME+ CONTACT + JSON_DATA);
+
+        assertNotNull(indexPageRes);
+        assertNotNull(contactPageRes);
+
+        PerReplicable indexRepl = indexPageRes.adaptTo(PerReplicable.class);
+        PerReplicable contactRepl = contactPageRes.adaptTo(PerReplicable.class);
+        assertNotNull(indexRepl);
+        assertNotNull(contactRepl);
+
+        try {
+            replication.replicate(indexPageRes, true);
+            replication.replicate(contactPageRes, true);
+        } catch (Replication.ReplicationException e) {
+            fail(e.getMessage());
+        }
+        // make published pages exist
+        assertFileExists(STATIC_HOME+ INDEX + HTML);
+        assertFileExists(STATIC_HOME+ INDEX + JSON_DATA);
+        assertFileExists(STATIC_HOME+ CONTACT + HTML);
+        assertFileExists(STATIC_HOME+ CONTACT + JSON_DATA);
+
+        // make sure deactivation does not remove published sibling
+        deactivateResource(CONTACT, contactPageRes, replication);
+        assertFileDoesNotExist(STATIC_HOME+ CONTACT + JSON_DATA);
+        assertFileDoesNotExist(STATIC_HOME+ CONTACT + HTML);
+        assertFileExists(STATIC_HOME+ INDEX + HTML);
+        assertFileExists(STATIC_HOME+ INDEX + JSON_DATA);
+        
     }
 
     private void assertFileExists(String filePath){

--- a/slingjunit.parent/core/src/main/java/com/peregrine/slingjunit/localFS/LocalFSJTest.java
+++ b/slingjunit.parent/core/src/main/java/com/peregrine/slingjunit/localFS/LocalFSJTest.java
@@ -1,0 +1,111 @@
+package com.peregrine.slingjunit.localFS;
+
+import com.peregrine.adaption.PerReplicable;
+import com.peregrine.replication.Replication;
+import com.peregrine.slingjunit.ReplicationTestBase;
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.junit.annotations.SlingAnnotationsTestRunner;
+import org.apache.sling.junit.annotations.TestReference;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.io.File;
+import java.util.Calendar;
+import static com.peregrine.slingjunit.author.RemoteReplAuthorJTest.STELLA_PNG;
+import static org.apache.jackrabbit.JcrConstants.JCR_LASTMODIFIED;
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertFalse;
+
+@RunWith(SlingAnnotationsTestRunner.class)
+public class LocalFSJTest extends ReplicationTestBase {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private Calendar beforeTime;
+    @TestReference
+    private ResourceResolverFactory resolverFactory;
+    private ResourceResolver adminResourceResolver = null;
+
+//    NOTE: There may be up to 3 types of replication services registered: localFS, remote, local
+//    If possible, remove OSGI configs for unused replication service implementations; remote and local service, before running this test.
+//    Otherwise, since the following line is non-deterministic, it could inject any of the replication services configured.
+    @TestReference
+    private Replication replication;
+
+    private Resource stellaImgRes;
+    private static String STATIC_HOME = "./sling/staticreplication";
+
+    @Before
+    public void setup(){
+        try {
+            beforeTime = Calendar.getInstance();
+            adminResourceResolver = resolverFactory.getAdministrativeResourceResolver(null);
+            stellaImgRes = adminResourceResolver.getResource(STELLA_PNG);
+        } catch (LoginException e) {
+            fail(e.getMessage());
+        }
+    }
+
+    @After
+    public void cleanup(){
+        adminResourceResolver.close();
+        adminResourceResolver = null;
+    }
+
+    @Test
+    public void setupIsCorrect(){
+        assertNotNull(replication);
+        assertTrue("localFS".equals(replication.getName()));
+        assertNotNull(stellaImgRes);
+    }
+
+    @Test
+    public void staticreplicationExists() {
+        assertFileExists(STATIC_HOME);
+    }
+
+
+    @Test
+    public void replicateOneAsset() {
+        // assert test starts with the file unpublished
+        assertFileDoesNotExist(STATIC_HOME+STELLA_PNG);
+        PerReplicable stellaImgRepl = stellaImgRes.adaptTo(PerReplicable.class);
+        assertFalse(stellaImgRepl.isReplicated());
+        try {
+            // publish the file
+            replication.replicate(stellaImgRes, true);
+            assertFileExists(STATIC_HOME+STELLA_PNG);
+            assertTrue(stellaImgRepl.isReplicated());
+            // touch the asset modified date after publishing
+            stellaImgRepl.getModifiableProperties().put(JCR_LASTMODIFIED, Calendar.getInstance());
+            assertNotNull(stellaImgRepl.getReplicated());
+            // verify the resource reports as stale
+            assertTrue(stellaImgRepl.isStale());
+
+        } catch (Replication.ReplicationException e) {
+            fail(e.getMessage());
+        }
+        // unpublish the file
+        deactivateResource(STELLA_PNG, stellaImgRes, replication);
+        // assert the file is unpublished
+        assertFileDoesNotExist(STATIC_HOME+STELLA_PNG);
+        // verify the resource accurately reports publication status
+        stellaImgRepl = adminResourceResolver.getResource(STELLA_PNG).adaptTo(PerReplicable.class);
+        assertFalse(stellaImgRepl.isReplicated());
+        assertFalse(stellaImgRepl.isStale());
+    }
+
+    private void assertFileExists(String filePath){
+        File file = new File(filePath);
+        assertTrue(file.exists());
+    }
+
+    private void assertFileDoesNotExist(String filePath){
+        File file = new File(filePath);
+        assertFalse(file.exists());
+    }
+}

